src/core/test/base/src/main/scala/TestBase.scala

```diff
@@ -153,3 +160,58 @@ abstract class TestBase extends FunSuite with BeforeAndAfterEachTestData with Be
   }
 
 }
+
+trait RoundTripTestBase extends TestBase {
+
+  val dfRoundTrip: DataFrame
+
+  val reader: MLReadable[_]
+
+  val modelReader: MLReadable[_]
+
+  val stageRoundTrip: PipelineStage with MLWritable
+
+  val savePath: String = Files.createTempDirectory("SavedModels-").toString
+
+  private def testRoundTripHelper(path: String,
+                                  stage: PipelineStage with MLWritable,
+                                  reader: MLReadable[_], df: DataFrame): Unit = {
+    stage.write.overwrite().save(path)
+    val loadedStage = reader.load(path)
+    (stage, loadedStage) match {
+      case (e1: Estimator[_], e2: Estimator[_]) =>
+        assert(e1.fit(df).transform(df).collect() === e2.fit(df).transform(df).collect())
+      case (t1: Transformer, t2: Transformer) =>
+        assert(t1.transform(df).collect() === t2.transform(df).collect())
+      case _ => throw new IllegalArgumentException(s"$stage and $loadedStage do not have proper types")
+    }
+    ()
+  }
+
+  def testRoundTrip(ignoreEstimators: Boolean = false): Unit = {
+    val fitStage = stageRoundTrip match {
+      case stage: Estimator[_] =>
+        if (!ignoreEstimators) {
+          testRoundTripHelper(savePath + "/stage", stage, reader, dfRoundTrip)
+        }
+        stage.fit(dfRoundTrip).asInstanceOf[PipelineStage with MLWritable]
+      case stage: Transformer => stage
+      case s => throw new IllegalArgumentException(s"$s does not have correct type")
+    }
+    testRoundTripHelper(savePath + "/fitStage", fitStage, modelReader, dfRoundTrip)
+
+    val pipe = new Pipeline().setStages(Array(stageRoundTrip))
+    if (!ignoreEstimators) {
+      testRoundTripHelper(savePath + "/pipe", pipe, Pipeline, dfRoundTrip)
+    }
+    val fitPipe = pipe.fit(dfRoundTrip)
+    testRoundTripHelper(savePath + "/fitPipe", fitPipe, PipelineModel, dfRoundTrip)
+  }
+
+  override def afterAll(): Unit = {
```

Comment:
Why not use this? https://commons.apache.org/proper/commons-io/javadocs/api-release/index.html?org/apache/commons/io/package-summary.html